nics: Improve the dialog experience with new framework components

mvollmer opened the pull request:
Demo:  https://youtu.be/VSb6anWw-tc

This uses the obvious porcelain components for checkboxes, dropdown selects and radio buttons. At the same time, the dialog experience is improved in a number of ways:

 - Instead of being read-only with a tooltip, the mac field is now disabled with a helper text.  The readonly state is rare and communicates less well than the disabled state. Tooltips (that open on hover) are not mobile friendly and generally awkward. There is enough space in the dialog for a inline explanation.

 - The "interface type" is selected with radio buttons instead of a dropdown. There are only three types and there is enough space to show them all. Instead of a info popover with questionably typography and content, the choices now have brief helper text to explain their purpose.

 - When a interface type has no sources, the type is disabled, instead of keeping it enabled and showing an empty list of sources, which would be a dead-end.

 - When editing a interface, the current source is always in the "Source" dropdown, even if it doesn't currently exist. Consequently, the current type is always enabled.

 - When adding a interface and there are no sources at all, the dialog explains that and only offers "Cancel".

<img width="845" height="606" alt="image" src="https://github.com/user-attachments/assets/3adf1890-caa8-487c-b75c-60d78a0b8c95" />

<img width="852" height="535" alt="image" src="https://github.com/user-attachments/assets/34fe3fa4-bae0-41d1-84c4-38b979639dae" />

<img width="1080" height="311" alt="image" src="https://github.com/user-attachments/assets/b07193cf-9e13-44d2-ab07-e12684562a8d" />

Fixes: [COCKPIT-1761](https://redhat.atlassian.net/browse/COCKPIT-1761)

[COCKPIT-1761]: https://redhat.atlassian.net/browse/COCKPIT-1761?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-26c86f56-20251202-155350-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-5168d18c-20251202-162035-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-dc7d51db-20251202-181824-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,910 @@
+        if (from_path in this.validation_state) {
+            console.log("RENAME", from_path, to_path);
+            this.validation_state[to_path] = this.validation_state[from_path];
+            delete this.validation_state[from_path];
```
These 7 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-ea2d88b2-20251203-143118-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,910 @@
+                const to = to_path + k.substring(from_path.length);
+                console.log("RENAME.", k, to);
+                this.validation_state[to] = this.validation_state[k];
+                delete this.validation_state[k];
```
These 6 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-ea2d88b2-20251203-143118-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 6 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-e22b04b7-20251204-081542-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,910 @@
+
+    static fromError(title: string, err: unknown) {
+        if (err && typeof err == "object" && "message" in err && typeof err.message == "string")
+            return new DialogError(title, err.message);
```
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-ea2d88b2-20251203-143118-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-ea2d88b2-20251203-143118-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,910 @@
+        if (Array.isArray(val)) {
+            return val.map((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
+        } else
+            return [];
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-9e650a26-20251203-151152-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,910 @@
+        const val = this.get();
+        if (Array.isArray(val)) {
+            for (let j = index; j < val.length - 1; j++)
+                this.dialog.rename_validation_state(this.path, j + 1, j);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-9e650a26-20251203-151152-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-9e650a26-20251203-151152-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,910 @@
+            await func(this.values);
+        } catch (ex) {
+            console.error(String(ex));
+            this.error = ex;
```
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-e22b04b7-20251204-081542-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,910 @@
+                if (this.online_validation)
+                    this.trigger_validation();
+                if (update_func)
+                    update_func(val);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-e22b04b7-20251204-081542-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,910 @@
+
+    constructor(title: string, details: React.ReactNode) {
+        this.title = title;
+        this.details = details;
```
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-e22b04b7-20251204-081542-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,910 @@
+    }
+
+    toString() {
+        return this.title + ": " + String(this.details);
```
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-e22b04b7-20251204-081542-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-e22b04b7-20251204-081542-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,918 @@
+        if (Array.isArray(val)) {
+            return val.map((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
+        } else
+            return [];
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-ec962d21-20251204-104905-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,918 @@
+        const val = this.get();
+        if (Array.isArray(val)) {
+            for (let j = index; j < val.length - 1; j++)
+                this.dialog.rename_validation_state(this.path, j + 1, j);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-ec962d21-20251204-104905-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-ec962d21-20251204-104905-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-dfd70a6b-20251208-133505-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,998 @@
+        const val = this.get();
+        if (Array.isArray(val)) {
+            for (let j = index; j < val.length - 1; j++)
+                this.dialog.rename_validation_state(this.path, j + 1, j);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e63f3e8-20251209-142520-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e63f3e8-20251209-142520-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,998 @@
+        if (err && typeof err == "object" && "message" in err && typeof err.message == "string")
+            return new DialogError(title, err.message);
+        else
+            return new DialogError(title, String(err));
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e63f3e8-20251209-142520-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/vm/nics/nicAdd.tsx:
```diff
@@ -73,275 +78,224 @@ function getRandomMac(vms: VM[]): string | undefined {
+                name="mac-setting"
+                isChecked={!v_set.get()}
+                label={_("Generate automatically")}
+                onChange={() => v_set.set(false)}
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e63f3e8-20251209-142520-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/vm/nics/nicAdd.tsx:
```diff
@@ -73,275 +78,224 @@ function getRandomMac(vms: VM[]): string | undefined {
-                    />
+    const defaultBody = (
+        <>
+            <Form onSubmit={e => e.preventDefault()} isHorizontal>
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e63f3e8-20251209-142520-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-dd682c79-20251209-152815-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-474d21ed-20251211-131002-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-336a2323-20251215-091435-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/vm/nics/nicAdd.tsx:
```diff
@@ -73,275 +79,224 @@ function getRandomMac(vms: VM[]): string | undefined {
+                name="mac-setting"
+                isChecked={!v_set.get()}
+                label={_("Generate automatically")}
+                onChange={() => v_set.set(false)}
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-336a2323-20251215-091435-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/vm/nics/nicAdd.tsx:
```diff
@@ -73,275 +79,224 @@ function getRandomMac(vms: VM[]): string | undefined {
+
+    const defaultBody = (
+        <>
+            <Form onSubmit={e => e.preventDefault()} isHorizontal>
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-336a2323-20251215-091435-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+            (val) => {
+                this.set(val);
+            },
+            this.#path,
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-1d0f9e39-20251215-171555-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+            this.#dialog._set_validation_timeout(
+                this.#path,
+                val,
+                debounce,
```
These 8 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-1d0f9e39-20251215-171555-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+                                .catch(
+                                    ex => {
+                                        console.error(ex);
+                                        return undefined;
```
These 7 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-1d0f9e39-20251215-171555-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+                                            console.log("DONE VALIDATE ASYNC", this.#path, result);
+                                            this.#dialog._set_validation_result(this.#path, val, result);
+                                        } else {
+                                            console.log("OUTDATED", this.#path);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-1d0f9e39-20251215-171555-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+                                        }
+                                    }
+                                );
+                    this.#dialog._set_validation_promise(this.#path, val, prom);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-1d0f9e39-20251215-171555-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+
+    _get_validation_state(path: string): DialogValidationState | null {
+        if (path in this.#validation_state)
+            return this.#validation_state[path];
```
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-1d0f9e39-20251215-171555-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-1d0f9e39-20251215-171555-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1378 @@
+        if (Array.isArray(val)) {
+            return val.map((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
+        } else
+            return [];
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-3b71159d-20251216-075539-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-6408b0c9-20251216-134315-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-3b71159d-20251216-075539-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1378 @@
+            (val, is_aux) => {
+                this.#setter(val, is_aux);
+            },
+            this.#path,
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-3b71159d-20251216-075539-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-6408b0c9-20251216-134315-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1378 @@
+            this.#dialog._set_validation_timeout(
+                this.#path,
+                val,
+                debounce,
```
These 8 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-3b71159d-20251216-075539-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 8 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-6408b0c9-20251216-134315-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-6408b0c9-20251216-134315-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1378 @@
+    }
+
+    set_aux(val: T): void {
+        this.#setter(val, true);
```
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-6408b0c9-20251216-134315-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1378 @@
+                                .catch(
+                                    ex => {
+                                        console.error(ex);
+                                        return undefined;
```
These 7 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-6408b0c9-20251216-134315-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1378 @@
+                                    result => {
+                                        if (this.#dialog._validation_state_is_current(this.#path, prom)) {
+                                            console.log("DONE VALIDATE ASYNC", this.#path, result);
+                                            this.#dialog._set_validation_result(this.#path, val, result);
```
These 5 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-6408b0c9-20251216-134315-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1378 @@
+                                            console.log("DONE VALIDATE ASYNC", this.#path, result);
+                                            this.#dialog._set_validation_result(this.#path, val, result);
+                                        } else {
+                                            console.log("OUTDATED", this.#path);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-6408b0c9-20251216-134315-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1378 @@
+                                        }
+                                    }
+                                );
+                    this.#dialog._set_validation_promise(this.#path, val, prom);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-6408b0c9-20251216-134315-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+
+function debug(...args: unknown[]) {
+    if (window.debugging == "all" || window.debugging?.includes("dialog"))
+        console.debug("dialog:", ...args);
```
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e45c3b5-20251217-171357-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-0ef97292-20251218-063518-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+    }
+
+    set_aux(val: T): void {
+        this.#setter(val, true);
```
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e45c3b5-20251217-171357-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-0ef97292-20251218-063518-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+        if (Array.isArray(val)) {
+            return val.map((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
+        } else
+            return [];
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e45c3b5-20251217-171357-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-0ef97292-20251218-063518-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+        const val = this.get();
+        if (Array.isArray(val)) {
+            for (let j = index; j < val.length - 1; j++)
+                this.#dialog._rename_validation_state(this.#path, j + 1, j);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e45c3b5-20251217-171357-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-0ef97292-20251218-063518-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+            this.#dialog,
+            () => this.get() as TT,
+            (val, is_aux) => {
+                this.#setter(val, is_aux);
```
These 7 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e45c3b5-20251217-171357-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 7 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-0ef97292-20251218-063518-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+            (val, is_aux) => {
+                this.#setter(val, is_aux);
+            },
+            this.#path,
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e45c3b5-20251217-171357-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-0ef97292-20251218-063518-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+
+    validate_async(debounce: number, func: (val: T) => Promise<string | undefined>): void {
+        const val = this.get();
+        this.#dialog._validate_value_async(this.#path, val, debounce, () => func(val));
```
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e45c3b5-20251217-171357-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+        if (state.timeout_id) {
+            debug("timeout cancel", state.path);
+            window.clearTimeout(state.timeout_id);
+            state.timeout_id = 0;
```
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e45c3b5-20251217-171357-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e45c3b5-20251217-171357-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+    }
+
+    #is_current_validation_round_id(id: unknown): boolean {
+        return Object.is(id, this.#validation);
```
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-0ef97292-20251218-063518-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-0ef97292-20251218-063518-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+        state: DialogValidationState,
+        val: unknown,
+        delay: number,
+        func: () => void,
```
These 5 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-0ef97292-20251218-063518-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-d596cf68-20251222-081408-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

mvollmer commented on test/check-machines-nics:
```diff
@@ -448,6 +448,32 @@ class TestMachinesNICs(machineslib.VirtualMachinesCase):
+        self.createVm("subVmTest1", running=False)
+
+        # Delete the "default" virtual network and hide all real interfaces.
+        m.execute("virsh net-destroy default; virsh net-undefine default")
```
There should be a cleanup that recreates the "default" network if it still missing.

Venefilyn reviewed (commented):
I'll take a look at the code tomorrow and try it locally

---

> Demo: https://youtu.be/VSb6anWw-tc

Good demo! Thanks!

> * The "interface type" is selected with radio buttons instead of a dropdown. There are only three types and there is enough space to show them all. Instead of a info popover with questionably typography and content, the choices now have brief helper text to explain their purpose.

So this is tricky UX wise. The general rule is that when you have 2 or 3 items it is better to have a radio button. When you get to 4 items or more it starts taking up quite a lot of space and can then be moved to a dropdown.

Overall, I like the changes you made but if we ever suspect to get another network interface type we should keep it as a dropdown like we have in main atm. There is a way to make the dropdown better, see Select example here where some items have a description, but adding links is probably not possible. But we can likely add the links and more information when you've select an interface type
https://www.patternfly.org/components/menus/select#select-option-variants

> * When adding a interface and there are no sources at all, the dialog offers to create the "default" network (which really should always exist anyway) in a preparation step. This means that the "network" type is always enabled.

This is great, good to see it also added within the network interfaces component as well.

Venefilyn commented on src/components/networks/networkList.tsx:
```diff
@@ -63,7 +79,28 @@ export class NetworkList extends React.Component<NetworkListProps> {
+                                                    >
+                                                        {_("Create the \"default\" virtual network")}
+                                                    </Button>
+                                                </EmptyStateFooter>
```
Split to a new component so we can reuse it in `nicsAdd` too. They're currently the same

Venefilyn commented on src/components/vm/nics/nicBody.tsx:
```diff
@@ -157,7 +157,18 @@ export function init_NetworkTypeAndSourceRow(
+            if (!avail.includes(src))
+                avail.push(src);
+        }
+    }
```
```suggestion
    // Make sure the current network source is available, no
    // matter what.
    if (network && network.type in _availableSourcesForType) {
        const avail = _availableSourcesForType[network.type];
        const src = getNetworkSource(network) || "";
        if (!avail.includes(src))
            avail.push(src);
    }
```

Venefilyn reviewed (changes requested):
Ping me if you have questions about OUIA. PF has `ouiaId` parameter for several components, otherwise you can use `data-ouia-component-id` directly on the DOM element.

Code overall seems fine, but would like to refactor one of the empty states. Also curious about the shell script we send to the backend - feels hardcoded and I wanna know if we won't get edge-cases where the IP would be wrong etc.

mvollmer commented on src/components/vm/nics/nicBody.tsx:
```diff
@@ -187,152 +205,84 @@ export function init_NetworkTypeAndSourceRow(
 
+export function NetworkTypeAndSource_has_sources(value: NetworkTypeAndSourceValue) {
+    return value.type != "";
+}
```
Yes, this is confusing!

mvollmer commented on src/components/vm/nics/vmNicsCard.tsx:
```diff
@@ -40,7 +40,8 @@ const getNetworkDevices = async (): Promise<Record<string, NetworkDevice>> => {
         const devs: Record<string, NetworkDevice> = {};
         for (const dev of output.trim().split('\n')) {
-            devs[dev] = {};
+            if (dev != "")
```
Roger.

mvollmer commented on test/check-machines-nics:
```diff
@@ -448,6 +448,32 @@ class TestMachinesNICs(machineslib.VirtualMachinesCase):
+        self.login_and_go("/machines")
+        self.waitPageInit()
+        self.goToVmPage("subVmTest1")
+        self.deleteIface(1)  # Remove default vNIC
```
Changed the type in machineslib.py

mvollmer committed:
nics: Improve dialogs with new framework components

 - Use DialogRadioSelect etc as appropriate.

 - Change TypeAndSource to disable types without sources instead of
   disabling the sources selector.

 - When editing a interface, always include the current source in the
   choices, even if it doesn't actually exist.

 - When adding and no sources exist, explain the situation without
   letting the user get stuck in a uncompletable dialog.

 - Editing interfaces with a unsupported type is now rejected with an
   error message.

mvollmer commented on src/components/vm/nics/nicAdd.tsx:
```diff
@@ -228,33 +234,55 @@ export const AddNIC = ({
-        <>
-            <Form onSubmit={e => e.preventDefault()} isHorizontal>
-                <NetworkTypeAndSourceRow field={dlg.field("type_and_source")} />
+    let defaultBody;
```
> Currently TS infers it to any due to it starting undefined and getting a value later on. 

TS infers different types in different places. After the `if`, `defaultBody` is `React.JSX.Element`.  Before it it is `any`, but since this gets narrowed down later, TS will not complain about it. We get errors for things that are truly `any`.

> but you need to add undefined to it

No, that is also not necessary. :-)  `ReactNode` already includes `undefined`, but even the following is fine:

```
    let foo: string;
    if (condition)
        foo = "xxx";
    else
        foo = "yyy";
```

`foo` does not need to be initialized and it's type does not need to include `undefined`. Everything is still sound since TS does not let you use a uninitialized variable.

So, IMO adding explicit types to things is good and let's do it here, but we can also rely on TS doing pretty clever inference. When there are no errors, the code is well typed and well checked, regardless of how many type annotations there are. (Because our config for tsc is strict.)

More type annotations are a help for future modifications of the code, so that mistakes are caught closer to their source, so we should have them when things are not obvious.

Venefilyn commented on src/components/vm/nics/nicBody.tsx:
```diff
@@ -45,78 +43,94 @@ export const NetworkModelRow = ({
     );
 };
 
+type NetworkType = "network" | "bridge" | "direct" | "user";
+
+export function isSupportedNetworkType(type: string): type is NetworkType {
+    return type == "network" || type == "bridge" || type == "direct" || type == "user";
+}
```
To "clean" this up so we don't need to duplicate stuff we can change it around a bit. Currently it doesn't type-check as well cause you could do `type == "ntework"` and it would not throw a type error

Following https://stackoverflow.com/a/51529486 we can do this instead, which although doesn't help us with the above makes it so we don't need to maintain both places and can just use `Array.includes()`
```ts
const networkType = ["network", "bridge", "direct", "user"] as const;
type NetworkType = (typeof networkType)[number];

export function isSupportedNetworkType(type: string): type is NetworkType {
    return networkType.includes((type as NetworkType))
}
```

Venefilyn commented on src/components/vm/nics/nicBody.tsx:
```diff
@@ -45,78 +43,94 @@ export const NetworkModelRow = ({
 }
 
-function getAvailableNetworkTypes(vm: VM, availableSources: AvailableSources) {
+interface AvailableSourcesForType {
+    network: string[];
+    bridge: string[];
+    direct: string[];
+    user: string[];
+}
```
```suggestion
type AvailableSourcesForType = {[key in NetworkType]: string[]}
```

mvollmer replied:
Yes, I learned about this just recently! But I don't mind being extra explicit here. If the set of supported network types is expanded, we have a lot of code churn anyway.

Venefilyn commented on src/components/vm/nics/nicBody.tsx:
```diff
@@ -45,78 +43,94 @@ export const NetworkModelRow = ({
-    detailParagraph?: React.ReactNode,
-    externalDocs?: React.ReactNode,
-    disabled?: boolean,
+    excuse: string | null,
```
Why `excuse: string|null` instead of `excuse?: string`

mvollmer replied:
No strong reason, just that this is enough for the current uses of this interface. If we make it optional, I would still  add the "| null" so we can write

```
    excuse: availableSources.bridge.length == 0 ? _("No bridge devices") : null
```

instead of

```
    ...(_availableSources.bridge.length == 0 ? { excuse:  _("No bridges") } : { })
````

which I find more cumbersome and less readable.

Venefilyn replied:
Would it not be possible to write `: undefined`?

Venefilyn reviewed (approved):
Overall LGTM I think! There are some type suggestions to reduce the duplication and human factor
